### DISCUSSION TITLE
Websocket chat example with room-specific subscription

### DISCUSSION THREAD

**Discussion by jodoox** [trpc/trpc#3369] [Q&A]
I've switched all my express endpoints to trpc over the last few weeks. It was a breeze, I'm really enjoying the experience so far.

I'm hitting a blocker as regards my Websocket updates: it is not clear from the docs how one should proceed for user/chatroom-specific updates, for instance: "John Doe is typing in room 'Star Wars'", only pushed to those that are active in the given room.

Should I just go for room-specific event names such as `ee.on('add.roomId', onAdd)` instead of  `ee.on('add', onAdd)` ? Or is there a better way?

Thanks

**Comment by KATT** [comment]
I think either you probably want viewer-specific events or room-specific

**Reply by jodoox** [reply]
As for my use case I'm looking for room-specific. What could be a good solution in your view?

**Comment by Elvincth** [comment]
Same question, any updates/methods you use? Thanks.

**Comment by gunhaxxor** [comment]
I've had the same challenge in the project I'm currently working in. I've ended up attaching events to my backend user instances. So in your case you could have an event(emitter) `someUserIsTypingInCurrentRoom` on every user. Then when subscription is setup/started the backend connects that user's event to the subscription observable.
Then you manually trigger the different events for the relevant users when needed. For example when a user is writing in a room the backend loops through the users in the room and trigger their respective `someUserIsTypingInCurrentRoom`. 

**Upsides:**
- The clients can start subscription for scoped (for example room specific) events without being in scope (being in a room). The backend keeps track of when to trigger the event. So in your case the users would need to subscribe to `someUserIsTypingInCurrentRoom` once and the subscription would "follow" them around to the different rooms.

**Downsides:**
- A lot of duplicated(ish) code where event(emitters) attached to users is more or less a duplication of the subscription endpoint (same payload and such). Changing the subscription api means also changing the events on the user object.
- Eventemitters are primarily aimed at broadcasting but in this case we are basically using each event as a 1 to 1 communication. I have a feeling this is not optimal 😏 

The last few days I've experimented with bypassing the eventemitters all together. I achieved it by doing this:
https://github.com/trpc/trpc/discussions/4032#discussion-4984609
I'm still not sure if that's a good idea but it seems to work fine.

I'm curious what other solutions people are using to this challenge!
